### PR TITLE
sort submission

### PR DESCRIPTION
**Pull request by hoangdinhtran1995**
This is a submission of the sort assignment in the algorithms and data structures class 2Q 2019.

Unfortunately could not understand the requirements for the median of medians assignment.

### CODE REVIEW

File: src/quick_select_improved.c [hunk 1]
```diff
@@ -18,22 +18,26 @@ A[0], A[1], ..., A[n-1] の中でk+1番目に小さい値を返す関数
...
-      swap(A+i, A+j);
-      j++;
+  for(i1 = j1 = 1; i1 < n; i1++){
+    if(A[i1] <= pivot){
+      swap(A+i1, A+j1);
+      j1++;
     }
   }
+  for(i2 = j2 = 0; i2 < j1; i2++){
+	  if(A[i2] <= pivot){
```
Comment (kondohyasuhiro) [inline src/quick_select_improved.c]: miss

File: src/quick_sort.c [hunk 2]
```diff
@@ -18,18 +18,44 @@ void swap(int *p, int *q){
 A[0], A[1], ..., A[n-1] をソートして昇順に書き換える関数
 */
 void quick_sort(int A[], int n){
```
Comment (kondohyasuhiro) [inline src/quick_sort.c]: If this program receive  `A = {3, ,3 ,3 }` , it will loop forever.

File: src/quick_sort_improved.c [hunk 3]
```diff
@@ -18,6 +18,31 @@ void swap(int *p, int *q){
...
+    }
+	
+    for(i2 = j2 = 1; i2 < j1; i2++){
+        if(A[i2] < pivot){
+            swap(A+i2, A+j2);
+            j2++;
+        }
+    }
+
+    quick_sort(A, j2);
```
Comment (kondohyasuhiro) [inline src/quick_sort_improved.c]: A[0] is still be the same value as  pivot.

**Comment by hoangdinhtran1995** [comment]
corrections done

**Comment by kondohyasuhiro** [comment]
Nice!

**Pull request closed without merging**